### PR TITLE
Make 2d segment segment distance more robust for kernel with inexact constructions

### LINKED ISSUE #4189: Wrong result of squared_distance of two Segment_2 objects
Wrong result of squared_distance of two Segment_2 objects.

## Issue Details

Using `CGAL::squared_distance` to compute the closest distance between two **almost collinear** segments may produce wrong results in some cases. e.g.
```
# Wavefront
v -4.0380854964382 -1.9947196614192 0.0
v 10.43442091460618 -0.5886833953492263 0.0
v -11.5138934277993 -2.721011070186227 0.0
v -8.822747585009402 -2.459560251317805 0.0
l 1 2
l 3 4
```
The computed result is 0, which is not right if we visualize the .obj file.

snapshot00

## Source Code

```cpp
#include <iostream>

#include <CGAL/Simple_cartesian.h>
#include <CGAL/squared_distance_2.h>

typedef CGAL::Simple_cartesian<double> Kernel;
typedef Kernel::Segment_2 Segment_2;

int main() {
  const Segment_2 segi = {
    { -4.0380854964382, -1.9947196614192 },
    { 10.43442091460618, -0.5886833953492263 } };
  const Segment_2 segj = {
    { -11.5138934277993, -2.721011070186227 },
    { -8.822747585009402, -2.459560251317805 } };

  const auto dist = CGAL::squared_distance(segi, segj);
  std::cout << "#dist: " << dist << std::endl;

  return 0;
}
```

## Environment

* Operating system (Windows/Mac/Linux, 32/64 bits): Windows and Linux, 64 bits
* Compiler: VC 2015, gcc 5.5.0
* Release or debug mode: Release
* Specific flags used (if any):
* CGAL version: 4.14
* Boost version: 1.65
* Other libraries versions if used (Eigen, TBB, etc.):

### PR DESCRIPTION
**Pull request by afabri**
## Summary of Changes

Avoiding subtractions fixes the issues when using `Simple_cartesian<double>`, and to use `orientation()` should make it robuster with `Epick`. 

The same should be done for Ray_2/Segment_2.   
We could introduce a predicate for 4 points that computes all orientation tests and determines at the same time which is the closer endpoint. This would make it yet robuster and it would avoid redundant computations.

I accidentally committed also a slightly faster implementation of  compare_signed_distance_to_line()`.

TODO:
- [x] Small feature

## Release Management

* Affected package(s): Distance_2
* Issue(s) solved (if any): fix #4189
* Small Feature: [CompareSignedDistanceToLine_2](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Kernel::CompareSignedDistanceToLine_2) I pre-approuve. -- Laurent Rineau 10:34, 12 March 2020 (CET)

image

### CODE REVIEW

**Comment by mglisse** [comment]
What is the impact on Epeck?

**Comment by afabri** [comment]
If I measure the time of 1mio distance computations for pairs of segments that go in some different branches the performance drops from 10.5 to 11 seconds.   
Ideally, we had a filtered predicate that computes the orientation of two points with respect to a segment and at the same time which of the two is  closer.  That would remove the re-computation of sub-determinants computed in the orientation test, and it would make it robust for further corner cases.  
As an alternative  we could also specialize for `Epick` and `Simple_cartesian<double>`.

**Comment by mglisse** [comment]
If the impact on Epeck (which is not the goal of this patch) is only about 5%, that's not so bad, I think further improvements can wait and don't need to block this.

File: Cartesian_kernel/include/CGAL/Cartesian/function_objects.h [hunk 1]
```diff
@@ -535,6 +535,40 @@ namespace CartesianKernelFunctors {
...
+    typedef Comparison_result   result_type;
+
+    result_type
+    operator()(const Point_2& a, const Point_2& b,
+               const Point_2& c, const Point_2& d) const
+    {
+      CGAL_kernel_precondition_code(Equal_2 equal;)
+      CGAL_kernel_precondition(! equal(a,b));
+      return cmp_signed_dist_to_lineC2( a.x(), a.y(),
+					b.x(), b.y(),
```
Comment (MaelRL) [inline Cartesian_kernel/include/CGAL/Cartesian/function_objects.h]: The indentation of the whole patch is pretty bad

File: Cartesian_kernel/include/CGAL/Cartesian/function_objects.h [hunk 2]
```diff
@@ -535,6 +535,40 @@ namespace CartesianKernelFunctors {
...
+					c.x(), c.y(),
+					d.x(), d.y());
+    }
+    /*
```
Comment (MaelRL) [inline Cartesian_kernel/include/CGAL/Cartesian/function_objects.h]: Why is this commented? If it's useless or broken, it should probably be removed.

File: Cartesian_kernel/include/CGAL/Cartesian/function_objects.h [hunk 3]
```diff
@@ -3893,7 +3927,7 @@ namespace CartesianKernelFunctors {
     }
   };
 
-  // TODO ...
+  // TODO:  Add functor Compare_signed_distance_to_line_2 and use this one in compare_signed_distance_to_line
```
Comment (MaelRL) [inline Cartesian_kernel/include/CGAL/Cartesian/function_objects.h]: Obsolete

File: Distance_2/include/CGAL/squared_distance_2_1.h [hunk 4]
```diff
@@ -197,7 +197,109 @@ namespace internal {
...
+	crossing2 = true;
+      } else {
+	crossing2 = (o2s == COLLINEAR);
+      }
+    }
+
+    c1s = CGAL::abs(wcross(seg2.source(), seg2.target(), seg1.source(), k));
+    c1e = CGAL::abs(wcross(seg2.source(), seg2.target(), seg1.target(), k));
+    c2s = CGAL::abs(wcross(seg1.source(), seg1.target(), seg2.source(), k));
+    c2e = CGAL::abs(wcross(seg1.source(), seg1.target(), seg2.target(), k));
```
Comment (MaelRL) [inline Distance_2/include/CGAL/squared_distance_2_1.h]: There is no need to compute all four values. For example, the first case only uses `c2s` and `c2e`.

File: Distance_2/include/CGAL/squared_distance_2_1.h [hunk 5]
```diff
@@ -197,7 +197,109 @@ namespace internal {
   typename K::FT
   squared_distance(const typename K::Segment_2 &seg1,
 		   const typename K::Segment_2 &seg2,
-		   const K& k)
+		   const K& k,
+                   const Cartesian_tag&)
```
Comment (MaelRL) [inline Distance_2/include/CGAL/squared_distance_2_1.h]: Is there still a need to disjoint Cartesian and Homogeneous kernels?

File: Homogeneous_kernel/include/CGAL/Homogeneous/function_objects.h [hunk 6]
```diff
@@ -770,6 +770,64 @@ namespace HomogeneousKernelFunctors {
     }
   };
 
+    template <typename K>
+  class Compare_signed_distance_to_line_2
```
Comment (MaelRL) [inline Homogeneous_kernel/include/CGAL/Homogeneous/function_objects.h]: Needs to be tested and documented in the Kernel concept... or just removed since it's not used by the PR.

File: Cartesian_kernel/include/CGAL/Cartesian/function_objects.h (hunk 2 above)
Comment (afabri) [reply Cartesian_kernel/include/CGAL/Cartesian/function_objects.h]: removed.

File: Cartesian_kernel/include/CGAL/Cartesian/function_objects.h (hunk 3 above)
Comment (afabri) [reply Cartesian_kernel/include/CGAL/Cartesian/function_objects.h]: removed

File: Distance_2/include/CGAL/squared_distance_2_1.h (hunk 4 above)
Comment (afabri) [reply Distance_2/include/CGAL/squared_distance_2_1.h]: changed.

File: Distance_2/include/CGAL/squared_distance_2_1.h (hunk 5 above)
Comment (afabri) [reply Distance_2/include/CGAL/squared_distance_2_1.h]: I don't want to work on Homogeneous.

File: Homogeneous_kernel/include/CGAL/Homogeneous/function_objects.h (hunk 6 above)
Comment (afabri) [reply Homogeneous_kernel/include/CGAL/Homogeneous/function_objects.h]: `Less_signed_distance_to_line_2`  now uses `Compare_..`.

File: Distance_2/include/CGAL/squared_distance_2_1.h (hunk 5 above)
Comment (MaelRL) [reply Distance_2/include/CGAL/squared_distance_2_1.h]: My point is that the code for the Cartesian kernel doesn't seem to be specific to a Cartesian kernel anymore (?)

**Comment by afabri** [comment]
@MaelRL what remains to be done?

**Comment by MaelRL** [comment]
https://github.com/CGAL/cgal/pull/4198#discussion_r319916927 (is there still a need to distinguish the Cartesian case?) and https://github.com/CGAL/cgal/pull/4198#discussion_r319917347 (must document Compare_signed_distance_to_line_2 in the Kernel concept/models) are still relevant.

**Comment by MaelRL** [comment]
For tracking, the current code does not fix the issue as the configuration reported in the issue still yields a distance of 0.

**Comment by afabri** [comment]
> 
> 
> For tracking, the current code does not fix the issue as the configuration reported in the issue still yields a distance of 0.

With VC++ it makes a difference. I added the issue.cpp to the tests, so I would be glad to see it in the testsuite

File: Distance_2/include/CGAL/squared_distance_2_1.h (hunk 5 above)
Comment (lrineau) [reply Distance_2/include/CGAL/squared_distance_2_1.h]: I agree with Mael, here: it looks like the the code with `Cartesian_tag` could be applied to the homogeneous case as well. I also think we could merge the two overloads of `squared_distance(Segment_2, Segment_2, K, Tag)` into one.

**Comment by afabri** [comment]
I had used the supposed generic code for the homogeneous kernel, but the testsuite "proves" that it does not work.    I prefer switching back to two different codes that work, instead of understanding what goes wrong for the homogeneous kernel, as I consider the homogeneous kernel as legacy.

**Comment by maxGimeno** [comment]
runtime errors in a Distance_2 test : https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-73.shtml#Distance_2

not entirely sure this PR is to blame as those are quite recent though...

File: Distance_2/test/Distance_2/issue_4189.cpp [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  return 0;
+}
+
+int main()
+{
+	fct<SC>();
+	fct<CGAL::Epick>();
+	fct<EC>();
+	return 0;
+}
```
Comment (lrineau) [inline Distance_2/test/Distance_2/issue_4189.cpp]: This file has several issues:
  1. It hardcodes `CGAL::Gmpq` unconditionally, and fails to compile on the few platforms where we have disabled GMP.
  2. It does not test anything: there are no `assert` to check that the results are the expected ones. It looks more like a debugging tool.

Comment (afabri) [reply Distance_2/test/Distance_2/issue_4189.cpp]: I changed it to `Exact_rational`.     The goal is not to assert whatsoever, but to see what the result is on different platforms.  It seems that on some platforms we will still see a really wrong distance.

Comment (lrineau) [reply Distance_2/test/Distance_2/issue_4189.cpp]: Then try to measure what it a "really wrong distance", and write the corresponding `assert`. Otherwise that test will be useless: we will just check that it compiles, but if, in a few years, suddenly the distance is wrong again, we will not notice.

**Comment by maxGimeno** [comment]
Travis is unhappy : https://travis-ci.org/CGAL/cgal/jobs/655368799#L2834

**Comment by afabri** [comment]
I find it rather annoying that `<CGAL/Filtered_kernel_selector.h>` includes the homogenous kernel.

File: Distance_2/test/Distance_2/issue_4189.cpp (hunk 7 above)
Comment (MaelRL) [reply Distance_2/test/Distance_2/issue_4189.cpp]: Is this resolved?

Comment (afabri) [reply Distance_2/test/Distance_2/issue_4189.cpp]: I just want to see  the difference but neither yellow nor red in the testsuite.

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-100.shtml

**Comment by sloriot** [comment]
Tested in CGAL-5.1-Ic-114

@afabri `CHANGE.md` not updated
